ecommerce/src/Relecloud.Web/Startup.cs

```diff
@@ -115,10 +116,81 @@ public void ConfigureServices(IServiceCollection services)
+                    {
+                        TransformRoleClaims(ctx);
+                        await CreateOrUpdateUserInformation(ctx);
+                    }
+                };
+            }));
+        }
+
+        private static async Task CreateOrUpdateUserInformation(TokenValidatedContext ctx)
+        {
+            try
+            {
+                if (ctx.Principal?.Identity is not null)
+                {
+                    // The user has signed in, ensure the information in the database is up-to-date.
+                    var user = new User
+                    {
+                        Id = ctx.Principal.GetUniqueId(),
+                        DisplayName = ctx.Principal.Identity.Name ?? "New User"
+                    };
+
+                    var repository = ctx.HttpContext.RequestServices.GetRequiredService<IConcertRepository>();
+                    await repository.CreateOrUpdateUserAsync(user);
+                }
+            }
+            catch (Exception ex)
+            {
+                var logger = ctx.HttpContext.RequestServices.GetRequiredService<ILogger<Startup>>();
+                logger.LogError(ex, "Unhandled exception from Startup.TransformRoleClaims");
+            }
+        }
+
+        private static void TransformRoleClaims(TokenValidatedContext ctx)
+        {
+            try
+            {
+                const string RoleClaim = "http://schemas.microsoft.com/ws/2008/06/identity/claims/role";
+                if (ctx.Principal?.Identity is not null)
+                {
+                    // Find all claims of the requested claim type, split their values by spaces
+                    // and then take the ones that aren't yet on the principal individually.
+                    var claims = ctx.Principal.FindAll("extension_AppRoles")
+                    .SelectMany(c => c.Value.Split(' ', StringSplitOptions.RemoveEmptyEntries))
+                    .Where(s => !ctx.Principal.HasClaim(RoleClaim, s)).ToList();
+
+                    // Add all new claims to the principal's identity.
+                    ((ClaimsIdentity)ctx.Principal.Identity).AddClaims(claims.Select(s => new Claim(RoleClaim, s)));
+                }
+            }
+            catch (Exception ex)
+            {
+                var logger = ctx.HttpContext.RequestServices.GetRequiredService<ILogger<Startup>>();
+                logger.LogError(ex, "Unhandled exception from Startup.TransformRoleClaims");
+            }
+        }
+
         // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
-        public void Configure(IApplicationBuilder app, IHostingEnvironment env, ApplicationInitializer applicationInitializer)
+        public void Configure(WebApplication app)
         {
-            if (env.IsDevelopment())
+            using var serviceScope = app.Services.CreateScope();
+            serviceScope.ServiceProvider.GetService<ApplicationInitializer>();
```

Comment:
What is this for?

Reply:
the code performs initialization on start. This code calls the initialization code on-start by requesting an object.

[ Initialize all resources at application startup.](https://github.com/Azure/fta-customerfacingapps/blob/2b95aa3593707dea7e3703673d285c6b88afcff2/ecommerce/src/Relecloud.Web/ApplicationInitializer.cs#L12)